Mathlib/Algebra/Algebra/Subalgebra/Tower.lean

```diff
@@ -0,0 +1,158 @@
+import Mathlib.Algebra.Algebra.Tower
+
+/-!
+# Subalgebras in towers of algebras
+
+In this file we prove facts about subalgebras in towers of algebra.
+
+An algebra tower A/S/R is expressed by having instances of `Algebra A S`,
+`Algebra R S`, `Algebra R A` and `IsScalarTower R S A`, the later asserting the
+compatibility condition `(r • s) • a = r • (s • a)`.
+
+## Main results
+
+ * `IsScalarTower.Subalgebra`: if `A/S/R` is a tower and `S₀` is a subalgebra
+   between `S` and `R`, then `A/S/S₀` is a tower
+ * `IsScalarTower.Subalgebra'`: if `A/S/R` is a tower and `S₀` is a subalgebra
+   between `S` and `R`, then `A/S₀/R` is a tower
+ * `Subalgebra.restrictScalars`: turn an `S`-subalgebra of `A` into an `R`-subalgebra of `A`,
+   given that `A/S/R` is a tower
+
+-/
+
+
+open Pointwise
+
+universe u v w u₁ v₁
+
+variable (R : Type u) (S : Type v) (A : Type w) (B : Type u₁) (M : Type v₁)
+
+namespace Algebra
+
+variable [CommSemiring R] [Semiring A] [Algebra R A]
+
+variable [AddCommMonoid M] [Module R M] [Module A M] [IsScalarTower R A M]
+
+variable {A}
+
+theorem lmul_algebraMap (x : R) :
+    LinearMap.Algebra.lmul R A (algebraMap R A x) = Algebra.lsmul R A x :=
+  Eq.symm <| LinearMap.ext <| smul_def x
+#align algebra.lmul_algebra_map Algebra.lmul_algebraMap
+
+end Algebra
+
+namespace IsScalarTower
+
+section Semiring
+
+variable [CommSemiring R] [CommSemiring S] [Semiring A]
+
+variable [Algebra R S] [Algebra S A]
+
+instance subalgebra (S₀ : Subalgebra R S) : IsScalarTower S₀ S A :=
+  of_algebraMap_eq fun _ ↦ rfl
+#align is_scalar_tower.subalgebra IsScalarTower.subalgebra
+
+variable [Algebra R A] [IsScalarTower R S A]
+
+instance subalgebra' (S₀ : Subalgebra R S) : IsScalarTower R S₀ A :=
+  @IsScalarTower.of_algebraMap_eq R S₀ A _ _ _ _ _ _ fun _ ↦
+    (IsScalarTower.algebraMap_apply R S A _ : _)
+#align is_scalar_tower.subalgebra' IsScalarTower.subalgebra'
+
+end Semiring
+
+end IsScalarTower
+
+namespace Subalgebra
+
+open IsScalarTower
+
+section Semiring
+
+variable {S A B} [CommSemiring R] [CommSemiring S] [Semiring A] [Semiring B]
+
+variable [Algebra R S] [Algebra S A] [Algebra R A] [Algebra S B] [Algebra R B]
+
+variable [IsScalarTower R S A] [IsScalarTower R S B]
+
+/-- Given a tower `A / ↥U / S / R` of algebras, where `U` is an `S`-subalgebra of `A`, reinterpret
+`U` as an `R`-subalgebra of `A`. -/
+def restrictScalars (U : Subalgebra S A) : Subalgebra R A :=
+  { U with
+    algebraMap_mem' := fun x ↦ by
+      rw [algebraMap_apply R S A]
+      exact U.algebraMap_mem _ }
+#align subalgebra.restrict_scalars Subalgebra.restrictScalars
+
+@[simp]
+theorem coe_restrictScalars {U : Subalgebra S A} : (restrictScalars R U : Set A) = (U : Set A) :=
+  rfl
+#align subalgebra.coe_restrict_scalars Subalgebra.coe_restrictScalars
+
+@[simp]
+theorem restrictScalars_top : restrictScalars R (⊤ : Subalgebra S A) = ⊤ :=
+  SetLike.coe_injective $ by dsimp -- porting note: why does `rfl` not work instead of `by dsimp`?
```

Comment:
The rest LGTM. Do you want to investigate further, or shall we merge this now and return to it later?

bors d+

Reply:
If you trace the `dsimp` it's using the `rfl` lemmas `coe_restrictScalars` and `Algebra.coe_top`. It's somewhat concerning that `rfl` doesn't time out (I wonder whether it gets to the kernel or not?), but on the other hand it's not so concerning that `rfl` lemmas help it finish the goal faster.

I think `dsimp` is a better proof since it's not relying on the definitions.

Reply:
I just let it run even longer and it did eventually timeout after almost 8 minutes with the expected message:
```text
(deterministic) timeout at 'whnf', maximum number of heartbeats (200000) has been reached (use 'set_option maxHeartbeats <num>' to set the limit)
```
Even though I agree that `by dsimp` is a better proof, I'm still very disappointed that `rfl` fails on something as simple as this, especially given that it works instantly in Lean 3.